src/query/service/src/sql/planner/binder/join.rs

```diff
@@ -448,9 +448,11 @@ impl<'a> JoinConditionResolver<'a> {
         // Bump types of left conditions and right conditions
         let left_type = left.data_type();
         let right_type = right.data_type();
-        let least_super_type = compare_coercion(&left_type, &right_type)?;
-        left = wrap_cast_if_needed(left, &least_super_type);
-        right = wrap_cast_if_needed(right, &least_super_type);
+        if left_type.ne(&right_type) {
```

Comment:
the `compare_coercion` already checked that `left_type` eq `right_type`.

I don't know the root reason why it works fine after your commit.